great_expectations_provider/hooks/external_connections.py

```diff
@@ -0,0 +1,452 @@
+"""
+External connection utilities for building database connection configurations
+from Airflow connections.
+"""
+
+import logging
+from pathlib import Path
+from typing import Literal, Optional, Union
+
+from airflow.hooks.base import BaseHook
+from airflow.models import Connection
+from pydantic import BaseModel, Field
+
+logger = logging.getLogger(__name__)
+
+
+class SnowflakeUriConnection(BaseModel):
+    """Pydantic model for URI-based Snowflake connection."""
+
+    type: Literal["uri"] = "uri"
+    connection_string: str
+
+
+class SnowflakeKeyConnection(BaseModel):
+    """Pydantic model for key-based Snowflake connection."""
+
+    model_config = {"extra": "forbid"}
+
+    type: Literal["key"] = "key"
+    account: str
+    user: str
+    role: Optional[str] = None
+    warehouse: str
+    database: str
+    schema_name: str = Field(..., alias="schema")
+    private_key: bytes
+
+
+def build_redshift_connection_string(conn_id: str, schema: Optional[str] = None) -> str:
+    """
+    Build connection string for Redshift connections.
+
+    Args:
+        conn_id: Airflow connection ID
+        schema: Optional schema override
+
+    Returns:
+        Connection string for Redshift
+
+    Raises:
+        ValueError: If connection doesn't exist or is invalid
+    """
+    conn = BaseHook.get_connection(conn_id)
+    if not conn:
+        raise ValueError(f"Connection does not exist in Airflow for conn_id: {conn_id}")
```

Comment:
```suggestion
        raise ValueError(f"Failed retrieving Airflow connection with conn_id: {conn_id}")
```

Similar suggestion across the other methods.